models/metrics/metrics.go

```diff
@@ -0,0 +1,132 @@
+	sent    metric.Int64Counter
+	bytes   metric.Int64Counter
+	latency metric.Int64Histogram
+}
+
+type promiseMetrics struct {
+	current   metric.Int64UpDownCounter
+	completed metric.Int64Counter
+}
+
+var (
+	events   eventMetrics
+	promises promiseMetrics
+)
+
+func metricName(name string) string {
+	return fmt.Sprintf("%s_%s", subsystem, name)
+}
+
+// Init initializes the arn sdk model metrics. This should only be called by the tattler constructor or tests.
+func Init(meter metric.Meter) error {
+	var err error
+	events.sent, err = meter.Int64Counter(metricName("event_sent_total"), metric.WithDescription("total number of events sent by the ARN client"))
+	if err != nil {
+		return err
+	}
+
+	events.bytes, err = meter.Int64Counter(metricName("event_sent_bytes_total"), metric.WithDescription("total number of bytes in event data sent by the ARN client"))
+	if err != nil {
+		return err
+	}
+
+	// TODO: adjust buckets
+	events.latency, err = meter.Int64Histogram(
+		metricName("event_sent_ms"),
+		metric.WithDescription("time spent to send ARN event"),
+		metric.WithExplicitBucketBoundaries(50, 100, 200, 400, 600, 800, 1000, 1250, 1500, 2000, 3000, 4000, 5000, 10000, 60000, 300000, 600000),
+	)
+	if err != nil {
+		return err
+	}
+
+	promises.completed, err = meter.Int64Counter(metricName("promise_total"), metric.WithDescription("total number of promises made by the ARN client"))
+	if err != nil {
+		return err
+	}
+
+	promises.current, err = meter.Int64UpDownCounter(metricName("current_promise_count"), metric.WithDescription("current number of promises made by the ARN client"))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SendEventSuccess increases the events.sent metric with success == true
+// and records the latency.
+func SendEventSuccess(ctx context.Context, elapsed time.Duration, inline bool, dataSize int64) {
+	opt := metric.WithAttributes(
+		attribute.Key(successLabel).Bool(true),
+		attribute.Key(inlineLabel).Bool(inline),
+	)
+	if events.sent != nil {
+		events.sent.Add(ctx, 1, opt)
+	}
+	if events.bytes != nil {
+		events.bytes.Add(ctx, dataSize, opt)
+	}
+	if events.latency != nil {
+		events.latency.Record(ctx, elapsed.Milliseconds(), opt)
+	}
+}
+
+// SendEventFailure increases the events.sent metric with success == false
+// and records the latency.
+func SendEventFailure(ctx context.Context, elapsed time.Duration, inline bool, dataSize int64) {
+	opt := metric.WithAttributes(
+		attribute.Key(successLabel).Bool(false),
+		attribute.Key(inlineLabel).Bool(inline),
+	)
+	if events.sent != nil {
+		events.sent.Add(ctx, 1, opt)
+	}
+	if events.bytes != nil {
+		events.bytes.Add(ctx, dataSize, opt)
+	}
+	if events.latency != nil {
+		events.latency.Record(ctx, elapsed.Milliseconds(), opt)
+	}
+}
+
+// Promise increases the promises.completed metric with timeout label.
+// This also decrements the current promise count.
+// This should be called on promise completion.
+func Promise(ctx context.Context, timeout bool) {
```

Comment:
I think this should be:

```go
func Promise(err error) {
  if err != nil {
    attribute.Key(errorLabel).String(error.Error())
  }
  ...
}
```

Then in msgs, this becomes:

```go
        if ctx.Err() != nil {
		metrics.Promise(ctx.Err())
		return ctx.Err()
	}

	select {
	case <-ctx.Done():
		metrics.Promise(models.ErrPromiseTimeout)
		return models.ErrPromiseTimeout
	case e := <-n.promise:
		metrics.Promise(e)
		return e
```

Reply:
Seems okay if the error messages don't get too long or numerous. Otherwise this could be high cardinality and have a large metric sample size.